tortoisebeacon/weakcoin/weak_coin.go

```diff
@@ -4,290 +4,303 @@ import (
+	smallestProposal []byte
+	allowances       UnitAllowances
+	// nextRoundBuffer is used to optimistically buffer messages from the next round.
+	nextRoundBuffer []Message
+	coins           map[types.RoundID]bool
+}
+
+// Get the result of the coinflip in this round. It is only valid inbetween StartEpoch/EndEpoch
+// and only after CompleteRound was called.
+func (wc *WeakCoin) Get(epoch types.EpochID, round types.RoundID) bool {
 	if epoch.IsGenesis() {
 		return false
 	}
 
-	pair := epochRoundPair{
-		EpochID: epoch,
-		Round:   round,
+	wc.mu.RLock()
+	defer wc.mu.RUnlock()
+	if wc.epoch != epoch {
+		wc.logger.Panic("requested epoch wasn't started or already completed",
+			log.Uint32("started_epoch", uint32(wc.epoch)),
+			log.Uint32("requested_epoch", uint32(epoch)))
 	}
 
-	wc.weakCoinsMu.RLock()
-	defer wc.weakCoinsMu.RUnlock()
+	return wc.coins[round]
+}
 
-	return wc.weakCoins[pair]
+// StartEpoch notifies that epoch is started and we can accept messages for this epoch.
+func (wc *WeakCoin) StartEpoch(epoch types.EpochID, allowances UnitAllowances) {
+	wc.mu.Lock()
+	defer wc.mu.Unlock()
+	wc.epoch = epoch
+	wc.allowances = allowances
 }
 
-func (wc *weakCoin) PublishProposal(ctx context.Context, epoch types.EpochID, round types.RoundID) error {
-	wc.Log.With().Info("Publishing proposal",
-		log.Uint64("epoch_id", uint64(epoch)),
+// CompleteEpoch completes epoch. After it is called Get for this epoch will panic.
+func (wc *WeakCoin) CompleteEpoch() {
+	wc.mu.Lock()
+	defer wc.mu.Unlock()
+	wc.allowances = nil
+	wc.coins = map[types.RoundID]bool{}
+	wc.round = 0
+}
+
+// StartRound process any buffered messages for this round and broadcast our proposal.
+func (wc *WeakCoin) StartRound(ctx context.Context, round types.RoundID) error {
+	wc.mu.Lock()
+	wc.logger.With().Info("started round",
+		log.Uint32("epoch_id", uint32(wc.epoch)),
 		log.Uint64("round_id", uint64(round)))
+	wc.round = round
+	wc.smallestProposal = nil
+	epoch := wc.epoch
+	for i, msg := range wc.nextRoundBuffer {
+		if msg.Epoch != wc.epoch || msg.Round != wc.round {
+			continue
+		}
+		if err := wc.updateProposal(msg); err != nil {
+			wc.logger.With().Debug("received invalid proposal",
+				log.Err(err),
+				log.Uint32("epoch_id", uint32(msg.Epoch)),
+				log.Uint64("round_id", uint64(msg.Round)))
+		}
+		wc.nextRoundBuffer[i].Epoch = 0
+		wc.nextRoundBuffer[i].Round = 0
+		wc.nextRoundBuffer[i].Unit = 0
+		wc.nextRoundBuffer[i].Signature = nil
+	}
+	wc.nextRoundBuffer = wc.nextRoundBuffer[:0]
+	wc.mu.Unlock()
+	return wc.publishProposal(ctx, epoch, round)
+}
 
-	proposal, err := wc.generateProposal(epoch, round)
+func (wc *WeakCoin) updateProposal(message Message) error {
+	buf := wc.encodeProposal(message.Epoch, message.Round, message.Unit)
+	miner, err := wc.verifier.Extract(buf, message.Signature)
 	if err != nil {
-		return err
+		return fmt.Errorf("can't recover miner id from signature: %w", err)
 	}
+	allowed, exists := wc.allowances[string(miner.Bytes())]
```

Comment:
Could you please add new lines after error checks? So the pieces of logic would be split and be easier to read